source/common/formatter/substitution_formatter.cc

```diff
@@ -378,10 +378,38 @@ std::vector<FormatterProviderPtr> SubstitutionFormatParser::parse(const std::str
                                      : "";
         // Validate the input specifier here. The formatted string may be destined for a header, and
         // should not contain invalid characters {NUL, LR, CF}.
-        if (std::regex_search(args, getStartTimeNewlinePattern())) {
+        if (std::regex_search(args, getDateFormatNewlinePattern())) {
           throw EnvoyException("Invalid header configuration. Format string contains newline.");
         }
         formatters.emplace_back(FormatterProviderPtr{new StartTimeFormatter(args)});
+      } else if (absl::StartsWith(token, "DOWNSTREAM_PEER_CERT_V_START")) {
```

Comment:
It seems like all of this code is going to be identical for START_TIME, here, etc. Please factor all of this and the formatters below into a single implementation for when you push the full PR with docs, etc.

Reply:
I went ahead and de-duped this logic by converting the StartTimeFormatter to SystemTimeFormatter, and keeping several subclass implementations that get used (and unit tested) independently.